### PR TITLE
pkg/transport: reload TLS certificates for every use

### LINKED ISSUE #7576: Certificate Rotation
# Versions
etcd Version: 3.1.3
etcd Version SHA: 21fdcc6

clientv3 Version SHA: 350d0cd

# Issue
I would like to start using `--client-cert-auth` and `--peer-client-cert-auth` to control access to my etcd cluster, however for security reasons I need to rotate these certificates fairly frequently.

1. Assuming I change the certificate in the file etcd is reading from, is there a way to ask etcd to reload the file?

2. If not, how bad of an idea is it to be restarting the etcd instances in a cluster frequently? For example, in a 3 node cluster having each instance restart once an hour?

3. And finally, any pointers on where I should start looking if I wanted to contribute this functionality to etcd?

Thank you.

### PR DESCRIPTION
**Pull request by tgrosinger**
This changes the baseConfig used when creating tls Configs to utilize
the GetCertificate and GetClientCertificate functions to always reload
the certificates from disk whenever they are needed.

Always reloading the certificates allows changing the certificates via
an external process without interrupting etcd.

Fixes #7576

### CODE REVIEW

**Comment by tgrosinger** [comment]
Caveats: This requires Go 1.8.1

**Comment by gyuho** [comment]
Commit title should be `pkg/transport: reload ...`?

**Comment by xiang90** [comment]
So the existing TLS connections will not be terminated even if the cert is changed? Is this the expected behavior?

**Comment by tgrosinger** [comment]
That is correct. Because the TLS negotiation occurs only when the connection is opened, ongoing connection can continue.

**Comment by xiang90** [comment]
@tgrosinger 

yea. i know how tls works. my concerns is that users might expect the existing tls connections to be terminated when they switch key/cert. but i am not a security expert and do not know the common practice. it would be great if you can feed me any related info: like what other projects do online cert/key change. thanks.

**Comment by Spindel** [comment]
Not terminating existing connections is the usual case. Main reason to replace certs is local renewal.  In the more rare case of moving from one trust root to another, it's a conscious act, and restarting the service would be a good way to terminate all connections.

Not killing existing connections matches expectation (either automatic reload, or reloading on a SIGHUP.)

File: pkg/transport/listener.go [hunk 1]
```diff
@@ -162,16 +162,26 @@ func (info TLSInfo) baseConfig() (*tls.Config, error) {
...
-	cfg := &tls.Config{
-		Certificates: []tls.Certificate{*tlsCert},
-		MinVersion:   tls.VersionTLS12,
-		ServerName:   info.ServerName,
+	cfg.GetCertificate = func(clientHello *tls.ClientHelloInfo) (
+		*tls.Certificate, error) {
+
+		// Load the certificate from disk every time so when it is replaced we
+		// will be using the latest version
+		return tlsutil.NewCert(info.CertFile, info.KeyFile, info.parseFunc)
```
Comment (xiang90) [inline pkg/transport/listener.go]: have you benchmarked the overhead of doing this? 

if this brings a huge overhead, we probably need to do caching. (inotify + caching parsed result or simply pull based caching)

File: pkg/transport/listener.go [hunk 2]
```diff
@@ -162,16 +162,26 @@ func (info TLSInfo) baseConfig() (*tls.Config, error) {
...
+		ServerName: info.ServerName,
 	}
 
-	cfg := &tls.Config{
-		Certificates: []tls.Certificate{*tlsCert},
-		MinVersion:   tls.VersionTLS12,
-		ServerName:   info.ServerName,
+	cfg.GetCertificate = func(clientHello *tls.ClientHelloInfo) (
+		*tls.Certificate, error) {
+
```
Comment (xiang90) [inline pkg/transport/listener.go]: remove this extra empty line at 172

File: pkg/transport/listener.go [hunk 3]
```diff
@@ -162,16 +162,26 @@ func (info TLSInfo) baseConfig() (*tls.Config, error) {
...
+		// Load the certificate from disk every time so when it is replaced we
+		// will be using the latest version
+		return tlsutil.NewCert(info.CertFile, info.KeyFile, info.parseFunc)
+	}
+	cfg.GetClientCertificate = func(unused *tls.CertificateRequestInfo) (
+		*tls.Certificate, error) {
+
```
Comment (xiang90) [inline pkg/transport/listener.go]: remove this empty line.

File: pkg/transport/listener.go [hunk 4]
```diff
@@ -162,16 +162,26 @@ func (info TLSInfo) baseConfig() (*tls.Config, error) {
...
+		// Load the certificate from disk every time so when it is replaced we
+		// will be using the latest version
+		return tlsutil.NewCert(info.CertFile, info.KeyFile, info.parseFunc)
```
Comment (xiang90) [inline pkg/transport/listener.go]: also consider caching?

**Comment by xiang90** [comment]
@Spindel thanks for the explanation. make sense.

**Comment by tgrosinger** [comment]
For our use case we are running etcd in a Kubernetes pod, so we are using an in-memory [EmptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) volume to help with performance.

I agree that caching would be desirable, however this patch was meant to be as minimal as possible. It's something I wanted to share with the community for others to build off of, not necessarily to be merged into etcd proper as it currently stands.

**Comment by gyuho** [comment]
Ok, I can confirm that this works with 

1. good cert etcd cluster
2. now has bad etcd cluster
3. client connection to this bad cert etcd cluster fails with `"transport: x509: certificate has expired or is not yet valid"; please retry`

Test fails because we still need `Certificates: []tls.Certificate{*tlsCert}` for initial listener start.
We need integration test around this, anyway. So I will cherry-pick this patch and create a separate PR with tests (today or tomorrow).

Thanks!

**Comment by gyuho** [comment]
Closing in favor of https://github.com/coreos/etcd/pull/7829.

**Pull request closed without merging**